File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaDataType.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.sdk.core.ValueRank;
import org.eclipse.milo.opcua.stack.core.NodeIds;
import org.eclipse.milo.opcua.stack.core.OpcUaDataType;
import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.UaStructuredType;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.eclipse.milo.opcua.stack.core.types.structured.EnumValueType;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.types.common.UaStructureUtilities;

import java.util.ArrayList;
import java.util.List;

public class UaDataType extends UaDefinitionNode {
    private @Nullable OpcUaDataType valueType;

    public UaDataType(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName,
            boolean isAbstract)
    {
        super(nodeId,browseName,displayName,isAbstract);
        valueType = null;
    }

    @Override
    public NodeClass nodeClass() {
        return NodeClass.DataType;
    }

    public OpcUaDataType getValueType()
    {
        if (null != valueType) return valueType;

        if (isSubtypeOf(NodeIds.Boolean))
        {
            valueType = OpcUaDataType.Boolean;
        } else if (isSubtypeOf(NodeIds.SByte)) {
            valueType = OpcUaDataType.SByte;
        } else if (isSubtypeOf(NodeIds.Int16)) {
            valueType = OpcUaDataType.Int16;
        } else if (isSubtypeOf(NodeIds.UInt16)) {
            valueType = OpcUaDataType.UInt16;
        } else if (isSubtypeOf(NodeIds.Int32)) {
            valueType = OpcUaDataType.Int32;
        } else if (isSubtypeOf(NodeIds.UInt32)) {
            valueType = OpcUaDataType.UInt32;
        } else if (isSubtypeOf(NodeIds.Int64)) {
            valueType = OpcUaDataType.Int64;
        } else if (isSubtypeOf(NodeIds.UInt64)) {
            valueType = OpcUaDataType.UInt64;
        } else if (isSubtypeOf(NodeIds.Float)) {
            valueType = OpcUaDataType.Float;
        } else if (isSubtypeOf(NodeIds.Double)) {
            valueType = OpcUaDataType.Double;
        } else if (isSubtypeOf(NodeIds.String)) {
            valueType = OpcUaDataType.String;
        } else if (isSubtypeOf(NodeIds.DateTime)) {
            valueType = OpcUaDataType.DateTime;
        } else if (isSubtypeOf(NodeIds.Guid)) {
            valueType = OpcUaDataType.Guid;
        } else if (isSubtypeOf(NodeIds.ByteString)) {
            valueType = OpcUaDataType.ByteString;
        } else if (isSubtypeOf(NodeIds.XmlElement)) {
            valueType = OpcUaDataType.XmlElement;
        } else if (isSubtypeOf(NodeIds.NodeId)) {
            valueType = OpcUaDataType.NodeId;
        } else if (isSubtypeOf(NodeIds.ExpandedNodeId)) {
            valueType = OpcUaDataType.ExpandedNodeId;
        } else if (isSubtypeOf(NodeIds.StatusCode)) {
            valueType = OpcUaDataType.StatusCode;
        } else if (isSubtypeOf(NodeIds.QualifiedName)) {
            valueType = OpcUaDataType.QualifiedName;
        } else if (isSubtypeOf(NodeIds.LocalizedText)) {
            valueType = OpcUaDataType.LocalizedText;
        } else if (isSubtypeOf(NodeIds.DataValue)) {
            valueType = OpcUaDataType.DataValue;
        } else if (isSubtypeOf(NodeIds.DiagnosticInfo)) {
            valueType = OpcUaDataType.DiagnosticInfo;
        } else if (isSubtypeOf(NodeIds.Enumeration)) {
            valueType = OpcUaDataType.Int32;
        } else if (isSubtypeOf(NodeIds.Structure)) {
            valueType = OpcUaDataType.ExtensionObject;
        }

        return valueType;
    }
}

File: opcua-webservice-node/source/src/server/addressspace/nodes/UaDataType.ts
import { NodeClass } from "opcua-webapi";
import { DataTypeIds, UaLocalizedText, UaNodeId, UaVariantType } from "opcua-webapi-ts";
import { UaDefinitionNode } from "./UaDefinitionNode";

export class UaDataType extends UaDefinitionNode {
    private _valueType: UaVariantType | null;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText,
        isAbstract: boolean) {
        super(nodeId, browseName, displayName, isAbstract);
        this._valueType = null;
    }

    get nodeClass(): NodeClass {
        return NodeClass.DataType;
    }

    getValueType(): UaVariantType | null {
        if (this._valueType !== null) {
            return this._valueType;
        }

        if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Boolean))) {
            this._valueType = UaVariantType.Boolean;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.SByte))) {
            this._valueType = UaVariantType.SByte;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Byte))) {
            this._valueType = UaVariantType.Byte;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Int16))) {
            this._valueType = UaVariantType.Int16;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.UInt16))) {
            this._valueType = UaVariantType.UInt16;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Int32))) {
            this._valueType = UaVariantType.Int32;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.UInt32))) {
            this._valueType = UaVariantType.UInt32;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Int64))) {
            this._valueType = UaVariantType.Int64;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.UInt64))) {
            this._valueType = UaVariantType.UInt64;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Float))) {
            this._valueType = UaVariantType.Float;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Double))) {
            this._valueType = UaVariantType.Double;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.String))) {
            this._valueType = UaVariantType.String;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.DateTime))) {
            this._valueType = UaVariantType.DateTime;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Guid))) {
            this._valueType = UaVariantType.Guid;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.ByteString))) {
            this._valueType = UaVariantType.ByteString;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.XmlElement))) {
            this._valueType = UaVariantType.XmlElement;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.NodeId))) {
            this._valueType = UaVariantType.NodeId;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.ExpandedNodeId))) {
            this._valueType = UaVariantType.ExpandedNodeId;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.StatusCode))) {
            this._valueType = UaVariantType.StatusCode;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.QualifiedName))) {
            this._valueType = UaVariantType.QualifiedName;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.LocalizedText))) {
            this._valueType = UaVariantType.LocalizedText;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Enumeration))) {
            this._valueType = UaVariantType.Int32;
        } else if (this.isSubtypeOf(UaNodeId.from(DataTypeIds.Structure))) {
            this._valueType = UaVariantType.ExtensionObject;
        }

        return this._valueType;
    }
}